keps/sig-testing/2539-continuously-deploy-k8s-prow/README.md

```diff
@@ -0,0 +1,149 @@
+- Errors caused by prow upgrades are discovered by prow alerts in a timely manner.
+
+
+### Goals
+
+The proposed change switches from daily manual deployment to hourly automated deployment.
+
+#### Prow Users
+
+Shouldn’t see any change, prow breakage should be discovered by prow monitoring system and rollback will be performed. The chance of prow being break is almost identical to what we have today(Assume there are not more than a single breaking change every day).
+
+#### Prow Oncall
+
+- What’s Not Changed
+  - React to prow alerts and take actions.
+- What’s Changed
+  - No more manual inspecting prow healthiness.
+  - No more manual lgtm/approve/retest autobump PRs.
+  - No more manual Slack posting.
+
+
+### Non-Goals
+
+Change how prow is released.
+
+
+## Proposal
+
+Prow autobump PRs are automatically merged every hour, only on working hours of working days.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Breaking Changes in Prow
+Breaking changes in prow will require manual intervention. Currently prow isn’t able to handle these intelligently, as it was not designed with the mindset of API versions and thus kubernetes conversion webhook can not help coping with breaking changes among major APIs.
+One possible way of dealing with breaking changes, is:
+- Prow oncall inspects prow logs and breaking changes announcements once per week, and take actions based on deprecation warnings from prow logs and breaking changes from ANNOUNCEMENTS.md.
+- [Stretch Goal][Push left] Discover breaking changes, especially configs or flag changes in prow integration test (This requires prow integration test use the same set of deployment configs as prod)
+- [Optional] (This is not very reliable) Prow TLs inspects new PRs, manually identifies possible breaking changes and informs oncall for awareness. Either prow TL or oncall can take deeper look at the new PR and decide whether to take actions or  not.
+
+## Design Details
+
+#### Automated Merging of Prow Autobump PRs
+
+- Prow autobump job is already configured to run on work days only, change it to at least one hour apart, so that it doesn’t bump more frequently than one hour.
+- Tide blindly trusts PRs from the bot that does autobump, merging the PR as long as tests all pass. Flaky tests will be covered by auto-pushing of new autobump jobs later.
+
+This approach uses tide auto-merge feature, so that no need to worry about repo requirements such as need more than one approver etc.
+
+```
+<<[UNRESOLVED (spiffxp) ]>>
+Suggestion: how to keep slack reports on each automated bump.
+<<[/UNRESOLVED]>>
+```
+
+#### Roll Back Process
+
+When prow stopped functioning after a bump, prow oncall should:
+- Stop auto-deploying by commenting `/hold` on latest autobump PR.
+- Manually create rollback PR for rolling back to known good version.
+- Manually apply the changes from rollback PR.
+
+```
+<<[UNRESOLVED]>>
```

Comment:
I would to see communication plan as part of this, e.g how and when will you announce this change:
- downstream
- to the users of prow.k8s.io

We should include plank version in testgrid.k8s.io jobs (cc @MushuEE)

I think this shouldn't be subject to v1.21 release freeze dates, but we should have a plan for how to be respectful toward the test freeze -> release phase.

This should have alpha/beta/ga phases called out. Low frequency as part of alpha. Test grid plumbed as part of beta. High frequency and decoupled jobs as part of GA

Reply:
Not super familiar with k8s release timeline. I see that `Test Freeze - Wednesday, March 24`, `The deadline to submit an entry is Thursday, March 25, EOD Pacific Time` from latest announcement. So the official release date is probably some time after March 25, should we wait until the official release of 1.21.0?

Reply:
Have updated https://github.com/kubernetes/enhancements/pull/2553 with:
- announcement
- alpha/beta/ga phases